### PR TITLE
Associate Policy with Role

### LINKED ISSUE #2253: Add Role --> Policy relationship
Relationship: Role **contains a** Policy

Keep it as a single entity relationship to make things simple - each policy has multiple rules, and hence allows adding more rules.

DataSteward Role contains DataStewardRoleAccessControlPolicy which has N number of rules. 

When creating a role, auto-create an empty policy

Deliverables:
- Establish Role --> Policy relationship during creation of a role
- Create an empty Policy when new roles are created

cc: @harshach

### PR DESCRIPTION
**Pull request by mithmatt**
- When a new Role is created, automatically create an Access Control Policy with empty set of Rules
- Create strong relationship: Role --contains--> Policy
- Modify RoleResource REST APIs to return policy

### Describe your changes :

I worked on this to fix #2253

Creating DevOps Role...

Screen Shot 2022-01-17 at 10 51 21 PM

Get DevOps Role with Policy ...

Screen Shot 2022-01-17 at 10 51 56 PM

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have performed a self-review of my own. 
- [x] I have tagged my reviewers below.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

Backend: @sureshms @harshach

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/exception/CatalogGenericExceptionMapper.java [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
   @Override
   public Response toResponse(Throwable ex) {
+    LOG.debug(ex.getMessage());
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/exception/CatalogGenericExceptionMapper.java:43]: This is really helpful while debugging

File: ingestion-core/src/metadata/_version.py [hunk 2]
```diff
@@ -7,5 +7,5 @@
 
 from incremental import Version
 
-__version__ = Version("metadata", 0, 8, 0, dev=4)
+__version__ = Version("metadata", 0, 8, 0, dev=5)
```
Comment (harshach) [inline ingestion-core/src/metadata/_version.py:10]: this should be taken care by the workflow itself

Comment (mithmatt) [reply ingestion-core/src/metadata/_version.py:10]: I broke a bunch of tests with my changes in this PR -- will attempt to fix this soon

Comment (harshach) [reply ingestion-core/src/metadata/_version.py:10]: never mind, this needs to manually done with latest changes cc @akash-jain-10

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/teams/RoleResourceTest.java [hunk 3]
```diff
@@ -56,16 +59,27 @@ public RoleResourceTest() {
...
     create = create(test, 3).withDescription("description");
-    createAndCheckEntity(create, adminAuthHeaders());
+    createAndCheckRole(create, adminAuthHeaders());
 
     create = create(test, 4).withDisplayName("displayName").withDescription("description");
-    createAndCheckEntity(create, adminAuthHeaders());
+    createAndCheckRole(create, adminAuthHeaders());
+  }
+
+  private Role createAndCheckRole(CreateRole create, Map<String, String> authHeaders) throws IOException {
```
Comment (mithmatt) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/teams/RoleResourceTest.java:74]: This is the new test for this feature.

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EntityResourceTest.java [hunk 4]
```diff
@@ -195,32 +196,33 @@ public void setup(TestInfo test) throws URISyntaxException, IOException {
...
+    TEAM_OWNER1 = new EntityReference().withId(TEAM1.getId()).withType(Entity.TEAM);
 
     RoleResourceTest roleResourceTest = new RoleResourceTest();
     ROLE1 = createRole(roleResourceTest.create(test), adminAuthHeaders());
     ROLE_REF1 = new EntityReference().withId(ROLE1.getId()).withType("role");
 
     // Ensure that DefaultAuthorizer gets enough time to load policies before running tests.
     try {
-      Thread.sleep(8000);
+      Thread.sleep(12000);
```
Comment (mithmatt) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EntityResourceTest.java:225]: Messed up here -- needs fixing before merging this PR

**Comment by mithmatt** [comment]
This needs to be rebased after #2281 is merged

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EntityResourceTest.java (hunk 4 above)
Comment (mithmatt) [reply catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/EntityResourceTest.java]: #2181 fixes this

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/RoleRepository.java [hunk 5]
```diff
@@ -45,10 +53,30 @@ public RoleRepository(CollectionDAO dao) {
...
   }
 
+  private EntityReference getPolicyForRole(Role role) throws IOException {
+    List<String> result =
+        daoCollection
+            .relationshipDAO()
+            .findFrom(
+                role.getId().toString(),
+                Entity.ROLE,
+                Relationship.APPLIED_TO.ordinal(),
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/RoleRepository.java:67]: should this be "belongs_to" not sure if we need "applied_to" as another relationship

Comment (harshach) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/RoleRepository.java:67]: we don't have "belongs_to" more like "contains"? @mithmatt

Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/RoleRepository.java]: Updated to role -- contains --> policy

**Comment by harshach** [comment]
@mithmatt approving the PR so that you can land after the failing test fix.

**Review by harshach — APPROVED** [review]

**Comment by mithmatt** [comment]
> @mithmatt approving the PR so that you can land after the failing test fix.

Thanks. We may have a flaky test here - was not able to reproduce the failure locally.

**Comment by mithmatt** [comment]
Test fails on line 316 here (only on CI)

image

**Comment by mithmatt** [comment]
Tests are passing now. 
Fixed a time wait in webhook to resolve this.

Lesson learned twice with this PR: Thread.sleep() can become a major source of problem leading to hard to trace failing tests. Will need to bring in awaitality for tests on asynchronous methods. 

We have #1987 open for this

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/events/WebhookResourceTest.java [hunk 6]
```diff
@@ -271,7 +271,7 @@ public void validateWebhookEvents() throws HttpResponseException, InterruptedExc
     assertNotNull(callbackEvents.peek());
     List<ChangeEvent> actualEvents =
         getChangeEvents("*", "*", "*", callbackEvents.peek().getTimestamp(), adminAuthHeaders()).getData();
-    waitAndCheckForEvents(actualEvents, callbackEvents, 10, 100);
+    waitAndCheckForEvents(actualEvents, callbackEvents, 15, 250);
```
Comment (mithmatt) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/events/WebhookResourceTest.java:274]: The method that was failing the test due to timing issues

**Pull request merged**